datalad/runner/utils.py

```diff
@@ -97,3 +91,31 @@ def process(self, data: str) -> List[str]:
 
     def finish_processing(self) -> Optional[str]:
         return self.remaining_data
+
+
+class AssemblingDecoderMixIn:
+    """ Mix in to safely decode data the is delivered in parts
+
+    This class can be used to decode data that is partially delivered.
+    It detects partial encodings and stores the non-decoded data to
+    combine it with additional data, that is delivered later,  and
+    decodes the combined data.
+
+    Any un-decoded data is stored in the 'remaining_data'-attribute.
+    """
+    def __init__(self):
+        self.remaining_data = defaultdict(bytes)
+
+    def decode(self, fd: int, data: bytes, encoding: str) -> str:
+        assembled_data = self.remaining_data[fd] + data
+        try:
+            unicode_str = assembled_data.decode(encoding)
+            self.remaining_data[fd] = b''
+        except UnicodeDecodeError as e:
+            unicode_str = assembled_data[:e.start].decode(encoding)
+            self.remaining_data[fd] = assembled_data[e.start:]
+        return unicode_str
+
+    def __del__(self):
+        if any(self.remaining_data.values()):
+            logger.warning("unprocessed data in AssemblingDecoderMixIn")
```

Comment:
may be add some `.debug` level message with actual content (in bytes decoded as ascii?) to ease troubleshooting if we do run into such case(s)?

Reply:
Good suggestion, will do that